docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,200 @@
+be taken by the `kubelet` unless the grace period has been exceeded.  A hard eviction
+threshold has no grace period, and if observed, the `kubelet` will take immediate action
+to reclaim the associated starved resource.
+
+Initial support for the feature will focus on available memory, but other potential signals
+that may be added in the future include cpu load or available disk.
+
+If no eviction thresholds are specified, the `kubelet` will not do any eviction.
+
+To configure hard eviction thresholds, the following flag will be supported:
+
+```
+--eviction-hard="": Comma-delimited list of eviction expressions.  For example, 'memory.available<300Mi'
+```
+
+To configure soft eviction thresholds, the following flags will be supported:
+
+```
+--eviction-soft="": Comma-delimted list of eviction expressions.  For example, 'memory.available<500Mi'
+--eviction-soft-period="": Comma-delimited list of eviction periods for each soft eviction signal.  For example, 'memory.available=30s'
+```
+
+If no eviction period is configured for a corresponding soft eviction threshold,
+it will default to 30s for each soft eviction signal.
+
+An eviction expression is of the form:
+
+`<eviction-signal><operator><quantity>`
+
+* valid `eviction-signal` tokens are `memory.available` (future signals are anticipated `storage` or `cpu.load`)
+* valid `operator` tokens are `<` and `>`
+* valid `quantity` tokens must match the quantity representation used by kubernetes
+
+
+## Eviction Monitoring Interval
+
+The `kubelet` will support monitoring available compute resource usage at the same
+specified housekeeping interval as `cAdvisor` housekeeping (10s in v1.2 by default).
+
+## Eviction scenario
+
+Let's assume the operator started the `kubelet` with the following:
+
+```
+--eviction-hard="memory.available<100Mi"
+--eviction-soft="memory.available<300Mi"
+--eviction-soft-period="memory.available=30s"
+```
+
+The `kubelet` will run a sync loop that looks at the available memory
+on the node as reported from `cAdvisor` by calculating (capacity - workingSet).
+If available memory is observed to drop below 100Mi, the `kubelet` will immediately
+initiate eviction. If available memory is observed as falling below `300Mi`,
+it will record when that signal was observed internally in a cache.  If at the next
+sync, that criteria was no longer satisfied, the cache is cleared for that
+signal.  If that signal is observed as being satisfied for longer than the
+specified period, the `kubelet` will initiate eviction to attempt to
+reclaim the resource that has met its eviction threshold.
+
+## Eviction of Pods
+
+If an eviction threshold has been met, the `kubelet` will initiate the
+process of evicting pods until it has observed the signal has gone below
+its defined threshold.
+
+The eviction sequence works as follows:
+
+* for each monitoring interval, if eviction thresholds have been exceeded
```

Comment:
Correct.

The current prototype I have does the following:

```
// called every monitoring interval
func (m *manager) synchronize() {
  stats := summary_stats
  if podsPendingEvictionStillActive(...) {
    return
  }
  if checkIfEvictionThresholdsAreMet(...) {
     podToEvict := strategy.Select(...)
     // toggle state to block incoming pods on admission, etc.
  }
}
```

So basically, we do not find a new pod to evict until we have definitively observed the old pod being evicted.  It has pros/cons, but it integrated best with current kubelet sync loop architecture.  Basically, the `getPodsToSync()``` ticker every 1s is used to query the stability manager to find the pods that need to be evicted.

Anyway, this is a little bit in the weeds, we can discuss that on the implementation PR.  Major thing is that we observe a pod eviction is completed before evicting another.